File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        count = [0] * 3
        j = ans = 0
        for i, ch in enumerate(s):
            count[ord(ch) - ord('a')] += 1
            while count[0] and count[1] and count[2]:
                count[ord(s[j]) - ord('a')] -= 1
                j += 1
            ans += j
        return ans

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var numberOfSubstrings = function(s) {
  const count = [0, 0, 0];
  const a = 'a'.charCodeAt();
  let [l, r, ans] = [0, 0, 0];
  
  while (r < s.length) {
    count[s.charCodeAt(r++) - a]++;
    while (count[0] && count[1] && count[2]) 
      --count[s.charCodeAt(l++) - a];
    
    ans += l;
  }
  
  return ans;
};